diff --git a/include/stdlib.h b/include/stdlib.h
index 42ca8336..10fb9d42 100644
--- a/include/stdlib.h
+++ b/include/stdlib.h
@@ -148,6 +148,7 @@ int clearenv(void);
 #endif
 
 #ifdef _GNU_SOURCE
+/*
 int ptsname_r(int, char *, size_t);
 char *ecvt(double, int, int *, int *);
 char *fcvt(double, int, int *, int *);
@@ -156,6 +157,7 @@ struct __locale_struct;
 float strtof_l(const char *__restrict, char **__restrict, struct __locale_struct *);
 double strtod_l(const char *__restrict, char **__restrict, struct __locale_struct *);
 long double strtold_l(const char *__restrict, char **__restrict, struct __locale_struct *);
+*/
 #endif
 
 #if defined(_LARGEFILE64_SOURCE) || defined(_GNU_SOURCE)
